### PR TITLE
[CI] Create zip of ray `session_latest/logs` dir on test failure and upload to buildkite via `/artifact-mount`

### PR DESCRIPTION
**Pull request by jon-chuang**
## Why are these changes needed?
Creates a zip of session_latest dir with test name and timestamp upon python test failure. Writes to dir specified by env var `RAY_TEST_FAILURE_LOGS_DIR`. Noop if env var does not exist.

Downstream consumer (e.g. CI) can upload all created artifacts in this dir. Thereby, PR submitters can more easily debug their CI failures, especially if they can't repro locally.

Limitations:
- a conftest.py file importing the main ray conftest.py needs to be present in same dir as test. This presents a challenge for e.g. dashboard tests which are highly scattered

## Related issue number
https://github.com/ray-project/ray/issues/23746

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jon-chuang** [comment]
It appears as of the CI run https://buildkite.com/ray-project/ray-builders-pr/builds/29186#bc5924e7-b7be-4738-95d8-ff3d40262b57, the experiment was a success

**Comment by jon-chuang** [comment]
Seems like spilled_objects was also zipped into the artifact. Perhaps I should only zip the log dir?

**Comment by krfricke** [comment]
Hi @jon-chuang, can we get another test run showing the result artifact? I can only see the linked example with the spilled objects.

**Comment by jon-chuang** [comment]
Not too sure what is the logs dir for windows. I'm currently hard coding the zipped dir as `/tmp/ray/session_latest/logs`. We could detect if os is windows and change this path accordingly.

**Comment by jon-chuang** [comment]
@krfricke do you have any suggestions on what the archive dir for storing the zipped logs dirs should be for windows and mac respectively? Seems the buildkite agents will upload from `/tmp/artifacts:/artifact-mount` on Linux, but not on Windows and Mac.

**Comment by jon-chuang** [comment]
I've decided to just enable this feature for Linux.

Windows and Mac tests can come later if someone could kindly advise me on the right way to configure an artifact mount dir for these.

**Comment by jon-chuang** [comment]
@krfricke you can take a look here: https://buildkite.com/ray-project/ray-builders-pr/builds/29541#a395d7cc-b8cb-41d4-b164-55a694ad9e1a

**Comment by jon-chuang** [comment]
@krfricke I am quite satisfied with the coverage for now. You can see the coverage and example artifacts here: https://buildkite.com/ray-project/ray-builders-pr/builds/29627

**Comment by jon-chuang** [comment]
The coverage is around dashboard tests, select serve & rllib and ML tests, and almost all python core tests.

**Review by krfricke — COMMENTED** [review]
Generally good, but let's clean up the code a bit please. Also, in some tests over 100 artifacts are generated (e.g. https://buildkite.com/ray-project/ray-builders-pr/builds/29627#cfb09371-0834-4f11-98ec-27ccde14f870) but Buildkite only supports up to 100. This is probably more something for a follow-up PR, but we could group some tests if this limit is exceeded

File: python/ray/tests/conftest.py
Comment (jon-chuang) [reply python/ray/tests/conftest.py:703]: went with `rep.nodeid.replace(os.sep, "::")`

**Comment by jon-chuang** [comment]
> Also, in some tests over 100 artifacts are generated (e.g. https://buildkite.com/ray-project/ray-builders-pr/builds/29627#cfb09371-0834-4f11-98ec-27ccde14f870) but Buildkite only supports up to 100.

https://buildkite.com/ray-project/ray-builders-pr/builds/29627#16e0350a-6204-48f8-8483-7b6acb051d96 seems to show that it supports much more than 100 artifacts. 

In practice I don't think any limit above 100 is a concern as we are expecting at most 1-5 tests to fail at a time (unless there is a test outage in which case probably any set of logs could help with diagnosis).

**Review by krfricke — APPROVED** [review]
Thanks for the refactor, this looks good to me. Will need code owner approval from other teams before we can merge.
While we're waiting for that, can you merge latest master and kick off CI again to fix the failing Ray client tests?

**Review by sven1977 — APPROVED** [review]

**Review by clarkzinzow — APPROVED** [review]

**Review by stephanie-wang — APPROVED** [review]

**Review by fishbone — APPROVED** [review]

**Review by simon-mo — APPROVED** [review]

**Comment by jon-chuang** [comment]
Seems the tests are passing now.

**Pull request merged**
